pkg/core/bundle_list.go

```diff
@@ -25,29 +28,339 @@ func minInt(a, b int) int {
+		// watch for an interruption signal requested by caller
+		wg.Add(1)
+		go watchForInterrupts(settings.doneChannel, &wg, doneWithKeysChan, doneWithBundlesChan)
 	}
 
+	keysChan := make(chan []string)
+	keyErrorChan := make(chan error)
+	bundleErrorChan := make(chan error)
+
+	// listening to errors from slave goroutines
+	wg.Add(1)
+	go watchForErrors(keyErrorChan, bundleErrorChan, errorChan, doneWithKeysChan, doneWithBundlesChan, &wg)
+
+	// starting keys retrieval
+	wg.Add(1)
+	go fetchKeys(repo, store, settings, keysChan, keyErrorChan, doneWithKeysChan, &wg) // scan for key batches
+
+	// start bundle metadata retrieval
+	wg.Add(1)
+	go fetchBundles(repo, store, settings, keysChan, batchChan, bundleErrorChan, doneWithBundlesChan, &wg)
+
+	// cleanup internal signaling channels
+	wg2.Add(1)
+	go func(wg, wg2 *sync.WaitGroup, inputChans ...chan<- struct{}) {
+		defer wg2.Done()
+		wg.Wait()
+		for _, ch := range inputChans {
+			close(ch)
+		}
+	}(&wg, &wg2, doneWithKeysChan, doneWithBundlesChan)
+
+	// return at once. Caller may chose to wait on returned WaitGroup
+	return batchChan, errorChan, &wg2
+}
+
+// watchForInterrupts broadcasts a done signal to several output channels
+func watchForInterrupts(doneChan <-chan struct{}, wg *sync.WaitGroup, outputChans ...chan<- struct{}) {
+	defer func() {
+		wg.Done()
+	}()
+
+	if _, interrupt := <-doneChan; interrupt {
+		for _, outputChan := range outputChans {
+			outputChan <- struct{}{}
+		}
+	}
+}
+
+// watchForErrors listens on 2 independent errors channels and report back to a main error channel.
+//
+// NOTE: this one does not need signaling, since any interruption bubbles up as error.
+func watchForErrors(
+	errorChan1, errorChan2 <-chan error,
+	outputChan chan<- error,
+	doneChan1, doneChan2 chan<- struct{}, wg *sync.WaitGroup) {
+	defer func() {
+		close(outputChan)
+		wg.Done()
+	}()
+
+	var (
+		err                            error
+		errorChan1Open, errorChan2Open bool = true, true
+	)
+	for errorChan1Open || errorChan2Open { // wait on both channels to close
+		select {
+		case err, errorChan1Open = <-errorChan1:
+			if errorChan1Open {
+				outputChan <- err
+				doneChan2 <- struct{}{}
+			} else {
+				errorChan1 = nil // no more selected case on this one
+				break
+			}
+		case err, errorChan2Open = <-errorChan2:
+			if errorChan2Open {
+				doneChan1 <- struct{}{}
+				outputChan <- err
+			} else {
+				errorChan2 = nil // no more selected case on this one
+				break
+			}
+		}
+	}
+}
+
+// fetchBundles waits on a channel of key batches and outputs batches of descriptors corresponding to these keys
+func fetchBundles(repo string, store storage.Store, settings CoreSettings,
+	keysChan <-chan []string,
+	batchChan chan<- model.BundleDescriptors, errorChan chan<- error,
+	doneChan <-chan struct{}, wg *sync.WaitGroup) {
+	defer func() {
+		close(batchChan)
+		close(errorChan)
+		wg.Done()
+	}()
+
+	for {
+		select {
+		case <-doneChan:
+			errorChan <- ErrInterrupted
+			return
```

Comment:
nitpick or possibly stylistic elucidation:  what `return` here where elsewhere `break`?

why this could be stylistic is that, for instance, i use double quotes for human readable strings in scripting languages like js and py and single quotes for keyword-like values.  is this a similar idiosyncratic convention where `return` consistently connotes a specific meaning wrt. programming patterns?

Reply:
break would just get us out from the `select`, not the loop as intended.